### ISSUE TITLE
Unable to read file produced by Impala

### ISSUE THREAD

**Issue by wesm** [dask/fastparquet#74]
Here's a link to the file: https://www.dropbox.com/s/fazbrkegbzzb7az/airlines2.parq?dl=0, let me know after you've downloaded it so I can delete from my Dropbox folder

```
In [6]: fp.ParquetFile('airlines2.parq').to_pandas()
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-6-61849c4e0274> in <module>()
----> 1 fp.ParquetFile('airlines2.parq').to_pandas()

/home/wesm/anaconda3/envs/parquet-test/lib/python3.5/site-packages/fastparquet/api.py in to_pandas(self, columns, categories, filters, index)
    293                              for (name, v) in views.items()}
    294                     self.read_row_group(rg, columns, categories, infile=f,
--> 295                                         index=index, assign=parts)
    296                     start += rg.num_rows
    297         else:

/home/wesm/anaconda3/envs/parquet-test/lib/python3.5/site-packages/fastparquet/api.py in read_row_group(self, rg, columns, categories, infile, index, assign)
    151         core.read_row_group(
    152                 infile, rg, columns, categories, self.helper, self.cats,
--> 153                 self.selfmade, index=index, assign=assign)
    154         if ret:
    155             return df

/home/wesm/anaconda3/envs/parquet-test/lib/python3.5/site-packages/fastparquet/core.py in read_row_group(file, rg, columns, categories, schema_helper, cats, selfmade, index, assign)
    300         raise RuntimeError('Going with pre-allocation!')
    301     read_row_group_arrays(file, rg, columns, categories, schema_helper,
--> 302                           cats, selfmade, assign=assign)
    303 
    304     for cat in cats:

/home/wesm/anaconda3/envs/parquet-test/lib/python3.5/site-packages/fastparquet/core.py in read_row_group_arrays(file, rg, columns, categories, schema_helper, cats, selfmade, assign)
    289         read_col(column, schema_helper, file, use_cat=use,
    290                  selfmade=selfmade, assign=out[name],
--> 291                  catdef=out[name+'-catdef'] if use else None)
    292 
    293 

/home/wesm/anaconda3/envs/parquet-test/lib/python3.5/site-packages/fastparquet/core.py in read_col(column, schema_helper, infile, use_cat, grab_dict, selfmade, assign, catdef)
    248     for defi, rep, val, d in out:
    249         if d and not use_cat:
--> 250             cval = dic[val]
    251         elif do_convert:
    252             cval = convert(val, se)

IndexError: index 240 is out of bounds for axis 0 with size 155
```

**Comment by mrocklin** [comment]
@martindurant have you had a chance to download this file?

**Comment by martindurant** [comment]
I now have the file.

**Comment by martindurant** [comment]
This happens only in the security_delay column, and only after data page 4. I do not see anything different about this column, or any way in which the data page is different from those before.

The remaining bytes read after the definition levels does not appear to be enough to fill all the values.

**Comment by mrocklin** [comment]
@wesm is there any chance that the file was generated in some invalid way?

@martindurant is there a way to make fastparquet robust, or at least informative, when it comes to invalid inputs?

**Comment by martindurant** [comment]
Warning and skipping apparent bad data is possible, but in this case we suspect the data might actually be valid.

**Comment by wesm** [comment]
I generated the file with Impala 2.5.0. I could regenerate it with just-released Impala 2.8.0 in case there is an idiosyncratic issue or a bug that was fixed. Near as I can tell I'm able to read the file with Impala, parquet-mr, and parquet-cpp, but I haven't checked whether the data those libraries return matches the source CSV files
